checks/process.go

```diff
@@ -106,28 +93,111 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
 	p.lastCPUTime = cpuTimes[0]
 	p.lastRun = time.Now()
 
-	statsd.Client.Gauge("datadog.process.containers.host_count", totalContainers, []string{}, 1)
-	statsd.Client.Gauge("datadog.process.processes.host_count", totalProcs, []string{}, 1)
+	statsd.Client.Gauge("datadog.process.containers.host_count", float64(totalContainers), []string{}, 1)
+	statsd.Client.Gauge("datadog.process.processes.host_count", float64(totalProcs), []string{}, 1)
 	log.Debugf("collected processes in %s", time.Now().Sub(start))
 	return messages, nil
 }
 
+func createProcCtrMessages(
+	procsByCtr map[string][]*model.Process,
+	containers []*model.Container,
+	cfg *config.AgentConfig,
+	sysInfo *model.SystemInfo,
+	groupID int32,
+) ([]model.MessageBody, int, int) {
+	totalProcs, totalContainers := 0, 0
+	msgs := make([]*model.CollectorProc, 0)
+
+	// we first split non-container processes in chunks
+	if procsByCtr[emptyCtrID] != nil {
+		totalProcs += len(procsByCtr[emptyCtrID])
+	}
+	chunks := chunkProcesses(procsByCtr[emptyCtrID], cfg.MaxPerMessage)
+	for _, c := range chunks {
+		msgs = append(msgs, &model.CollectorProc{
+			HostName:  cfg.HostName,
+			Info:      sysInfo,
+			Processes: c,
+			GroupId:   groupID,
+		})
+	}
+
+	// pack all containered processes with containers in a message
+	validIndex := 0
+	ctrProcs := make([]*model.Process, 0)
+	for _, ctr := range containers {
+		// if all processes are skipped for this container, don't send the container
+		if _, ok := procsByCtr[ctr.Id]; !ok {
+			continue
+		}
+		totalProcs += len(procsByCtr[ctr.Id])
+		totalContainers++
+		ctrProcs = append(ctrProcs, procsByCtr[ctr.Id]...)
+
+		// modify in-place and increase index, keep track of valid containers
+		containers[validIndex] = ctr
+		validIndex++
+
+	}
+
+	if len(containers[:validIndex]) > 0 {
+		msgs = append(msgs, &model.CollectorProc{
+			HostName:   cfg.HostName,
+			Info:       sysInfo,
+			Processes:  ctrProcs,
```

Comment:
Isn't this going to be every containerized-process on the host? That's going to be quite large won't it? Could we write a loop like:


```go
var procContainerMessages []CollectorProc
nextMessage . = . CollectorProc{}
nextMessageSize = 0
for _, ctr := containers {
   // skip if no procs for ctr
  size :=  size_of_container + size_of_proc * procsByCtr[ctr.Id]
   if size + nextMessageSize > max {

         procContainerMessages = append(procContainerMessages, nextMessage)
         nextMessage = . {}
         nextMessageSize = 0
    } else {
         nextMessage.Containers = append(...)
         nextMesage.Processes = append(...)
    } 
}
```

and get rid of the previous loop altogether?

Reply:
I thought about this as well, but I want to simply the logic on the first run because I think the chance that a host runs 100+ containers is small(our chunk size is 100). But since you wrote the logic and it doesn't seem to be very complex, I might give it a shot.

Reply:
Per in-person convo with Lee I'll leave the current logic in place.